futures-util/src/future/shared.rs

```diff
@@ -156,153 +203,76 @@ impl<F> Future for Shared<F>
                 }
             }
 
-            let _reset = Reset(&self.inner.notifier.state);
+            let _reset = Reset(&this.inner.notifier.state);
 
             // Poll the future
             let res = unsafe {
-                let (ref mut f, ref mut data) = *(*self.inner.future.get()).as_mut().unwrap();
-                let waker = Waker::from(self.inner.notifier.clone());
-                let mut cx = task::Context::new(data, &waker, cx.executor());
-                f.poll(&mut cx)
+                if let FutureOrOutput::Future(future) = &mut *this.inner.future_or_output.get() {
+                    PinMut::new_unchecked(future).poll(&mut cx)
```

Comment:
I suggest to put the call to `poll` outside of the unsafe block

Reply:
It gets messy here because it's prefixed by the call to `new_unchecked`, which uses a temporary `&mut` created by the `&mut *` on `UnsafeCell::get`, so getting the lifetimes right so that we don't create a temporary `&mut` which is alive until the following `get` on line 234 requires an extra set of braces. I'd prefer to leave it as-is :laughing:

Reply:
Yes definitely! This makes sense